### DISCUSSION TITLE
Trying to narrow down Telegram odds API bot

### DISCUSSION THREAD

**Discussion by CryptoZealott** [cloudbet/docs#65] [Q&A]
Hello all, I am trying to finish up my telegram bot script that requests the odds for a certain sports league and posts them in a telegram channel when prompted with a command. However as it stands it is pulling the odds for every event in the certain sport, rather than just the certain sports league.

For example I want it to pull the ice_hockey.winner odds for only NHL games, and not every ice hockey game on Cloudbet. Any help or direction would be greatly appreciated. Code below:

`    from datetime import datetime
    from pprint import pprint
    import time
    import pytz
    import requests
    import telepot
    from telepot.loop import MessageLoop

    # Define the Telegram bot token and chat ID
    bot_token = 'xxxx'
    chat_id = 'xxxx'

    # create a requests session with the authentication header
    session = requests.Session()
    session.headers["X-API-Key"] = 'xxxx'

    # Define function for sending the message to Telegram
        def send_telegram_message(message):
            bot = telepot.Bot(bot_token)
            bot.sendMessage(chat_id, message)

    # Define function for scraping Cloudbet.com for odds
        def scrape_cloudbet():
    # get today's date formatted as yyyy-mm-dd
    date = datetime.utcnow().strftime("%Y-%m-%d")

    # perform the request
    url = "https://sports-api.cloudbet.com/pub/v2/odds/fixtures"
    response = session.get(url, params={"sport": "ice_hockey", "date": date, "limit": "12"})

    # extract the events data
    response_data = response.json()
    events = []
    for competition in response_data["competitions"]:
        competition_name = competition["category"]["name"] + " - " + competition["name"]
        for event in competition["events"]:
            events.append({
                "id": event["id"],
                "name": event["name"],
                "competition": competition_name,
            })

    # extract the selections data
    selections = []
    for event in events:
        url = f"https://sports-api.cloudbet.com/pub/v2/odds/events/{event['id']}"
        response = session.get(url, params={"markets": "ice_hockey.winner"})
        response_data = response.json()
        event_selections = (
            response_data
            .get("markets", {})
            .get("ice_hockey.winner", {})
            .get("submarkets", {})
            .get("period=ot\u0026period=ft", {})
            .get("selections", [])
        )
        for event_selection in event_selections:
            selections.append({
                "event_id": event["id"],
                "event": event["name"],
                "competition": event["competition"],
                "home": response_data["home"]["name"],
                "away": response_data["away"]["name"],
                "outcome": event_selection["outcome"],
                "price": event_selection["price"],
                "status": event_selection["status"],
            })

    # format the selections data and send it to Telegram
    selections_message = "\n".join([f"{selection['event']} - {selection['competition']}: {selection['home']} vs {selection['away']}, {selection['outcome']} @ {selection['price']}" for selection in selections])
    message = f"Here are the latest NHL odds:\n{selections_message}"
    send_telegram_message(message)

    # Define function to handle Telegram commands
    def handle_telegram_command(msg):
        chat_id = msg['chat']['id']
        command = msg['text']
        if command == '/odds':
            scrape_cloudbet()

    # Set up Telegram message loop and start listening for commands
    bot = telepot.Bot(bot_token)
    MessageLoop(bot, handle_telegram_command).run_forever()`

**Comment by kgravenreuth** [comment]
Hi @CryptoZealott ,
Your telegram bot sounds like an interesting project!

If you are only interested in NHL events, then we would recommend you to:

1. Get the NHL competition key from the `/v2/odds/sports/{key}` endpoint: 

```
curl -X 'GET' \
  'https://sports-api.cloudbet.com/pub/v2/odds/sports/ice_hockey' \
  -H 'accept: application/json' \
  -H 'X-API-Key: <YOUR_API_KEY>'
```

2. Then query for all events of the NHL competition from the `/v2/odds/competitions/{key}` endpoint for a given timerange: 

```
curl -X 'GET' \
  'https://sports-api.cloudbet.com/pub/v2/odds/competitions/ice-hockey-usa-nhl?from=1686240000&to=1686326399&limit=12' \
  -H 'accept: application/json' \
  -H 'X-API-Key: <YOUR_API_KEY>'
```

Please let us know if this helps, thank you.

Regards,
Klaus Gravenreuth
Product & Engineering @ Cloudbet

**Reply by CryptoZealott** [reply]
thanks Klaus, Im gonna work on tweaking this today

**Comment by Charlotte-br560** [comment]
It seems like you're pulling odds for all ice hockey events instead of just NHL games. To filter for NHL games, you need to modify your code to include a condition that checks the league for each event before processing its odds. Here's a suggestion:

```
from datetime import datetime
import requests
import telepot
from telepot.loop import MessageLoop

# Define the Telegram bot token and chat ID
bot_token = 'xxxx'
chat_id = 'xxxx'

# Define function for sending the message to Telegram
def send_telegram_message(message):
    bot = telepot.Bot(bot_token)
    bot.sendMessage(chat_id, message)

# Define function for scraping Cloudbet.com for odds
def scrape_cloudbet():
    # Get today's date formatted as yyyy-mm-dd
    date = datetime.utcnow().strftime("%Y-%m-%d")
    
    # Perform the request
    url = "https://sports-api.cloudbet.com/pub/v2/odds/fixtures"
    response = requests.get(url, params={"sport": "ice_hockey", "date": date, "limit": "12"})
    
    # Extract NHL events and their odds
    selections = []
    response_data = response.json()
    for competition in response_data["competitions"]:
        league_name = competition["category"]["name"]
        if league_name == "NHL":
            for event in competition["events"]:
                event_name = event["name"]
                home_team = event["home"]["name"]
                away_team = event["away"]["name"]
                event_id = event["id"]
                odds = get_odds_for_event(event_id)
                if odds:
                    selections.append(f"{event_name}: {home_team} vs {away_team}, Odds: {odds}")

    # Send NHL odds to Telegram
    if selections:
        message = "Here are the latest NHL odds:\n" + "\n".join(selections)
        send_telegram_message(message)
    else:
        send_telegram_message("No NHL games found for today.")

# Define function to get odds for a specific event
def get_odds_for_event(event_id):
    url = f"https://sports-api.cloudbet.com/pub/v2/odds/events/{event_id}"
    response = requests.get(url, params={"markets": "ice_hockey.winner"})
    response_data = response.json()
    if "markets" in response_data and "ice_hockey.winner" in response_data["markets"]:
        odds = response_data["markets"]["ice_hockey.winner"]["submarkets"]["period=ot\u0026period=ft"]["selections"]
        return ", ".join([f"{outcome['outcome']}: {outcome['price']}" for outcome in odds])
    else:
        return None

# Define function to handle Telegram commands
def handle_telegram_command(msg):
    command = msg['text']
    if command == '/odds':
        scrape_cloudbet()

# Set up Telegram message loop and start listening for commands
bot = telepot.Bot(bot_token)
MessageLoop(bot, handle_telegram_command).run_forever()

```

**Comment by Emmeline-Hawthorne** [comment]
Looks like you're pulling all ice hockey events instead of filtering for just NHL. Try adding a check in your loop to only include events where competition["name"] matches "NHL". Something like this:
```
`for competition in response_data["competitions"]:
    if competition["name"] == "NHL":  # Filter only NHL events
        competition_name = competition["category"]["name"] + " - " + competition["name"]
        for event in competition["events"]:
            events.append({
                "id": event["id"],
                "name": event["name"],
                "competition": competition_name,
            })
````
